framework/test/jest/unit/specs/modules/chain/synchronizer/synchronizer.spec.js

```diff
@@ -174,86 +434,104 @@ describe('Synchronizer', () => {
-			await expect(synchronizer.run({ id: 'a blockId' })).rejects.toThrow(
+			await expect(synchronizer.run(aReceivedBlock, aPeerId)).rejects.toThrow(
 				'Synchronizer is already running',
 			);
 		});
 
-		it('should verify the block before sync', async () => {
-			await synchronizer.run(receivedBlock);
+		it('should reject with error if required properties are missing (block)', async () => {
+			await expect(synchronizer.run()).rejects.toThrow(
+				'A block must be provided to the Synchronizer in order to run',
+			);
+			expect(synchronizer.active).toBeFalsy();
+		});
 
-			expect(processorMock.validateDetached).toHaveBeenCalledTimes(1);
-			expect(processorMock.validateDetached).toHaveBeenCalledWith(
-				receivedBlock,
+		it('should reject with error if required properties are missing (peerId)', async () => {
+			await expect(synchronizer.run({ height: 1 })).rejects.toThrow(
+				'A peer ID from the peer sending the block must be provided to the Synchronizer in order to run',
 			);
+			expect(synchronizer.active).toBeFalsy();
 		});
 
-		it('should reject with error if block verification failed', async () => {
-			const validationError = new Error('Block verifyError');
-			processorMock.validateDetached.mockRejectedValue(validationError);
+		it('should validate the block before sync', async () => {
+			jest.spyOn(processorModule, 'validateDetached');
 
-			await expect(synchronizer.run({ id: '1234' })).rejects.toThrow(
-				validationError,
+			await synchronizer.run(aReceivedBlock, aPeerId);
+
+			expect(processorModule.validateDetached).toHaveBeenCalledWith(
+				await processorModule.deserialize(aReceivedBlock),
 			);
 		});
 
-		it('should determine the sync mechanism for received block', async () => {
-			await synchronizer.run(receivedBlock);
+		it('should reject with error if block validation failed', async () => {
+			try {
+				await synchronizer.run(
+					{
+						...aReceivedBlock,
+						blockSignature: '12312334534536645656',
+					},
+					aPeerId,
+				);
+			} catch (error) {
+				expect(error[0].message).toEqual('should match format "signature"');
+				expect(synchronizer.active).toBeFalsy();
+			}
+		});
+
+		it('should determine the sync mechanism for received block and run it', async () => {
+			syncMechanism1.isValidFor.mockResolvedValue(true);
+			syncMechanism2.isValidFor.mockResolvedValue(false);
 
-			expect(synchronizer._determineSyncMechanism).toHaveBeenCalledTimes(1);
-			expect(synchronizer._determineSyncMechanism).toHaveBeenCalledWith(
-				receivedBlock,
+			await synchronizer.run(aReceivedBlock, aPeerId);
+
+			expect(syncMechanism1.isValidFor).toHaveBeenCalledTimes(1);
+			expect(syncMechanism2.isValidFor).toHaveBeenCalledTimes(1);
+			expect(syncMechanism1.run).toHaveBeenCalledWith(
+				await processorModule.deserialize(aReceivedBlock),
+				aPeerId,
 			);
+			expect(syncMechanism2.run).not.toHaveBeenCalled();
+			expect(loggerMock.info).nthCalledWith(2, 'Triggering: Object');
```

Comment:
I prefer to avoid checking logs, is there any other way to check this behavior?

Reply:
I am also not a fan of checking the logs.. but for this scenario, it should be fine.
maybe we can use setter and getter to get the mechanism? just a thought

Reply:
I think logs are important especially for delegates, and if the info we are dumping to the logs is important we should have tests making sure we are still doing that.